toolchain/check/import_ref.cpp

```diff
@@ -125,44 +125,58 @@ auto VerifySameCanonicalImportIRInst(Context& context, SemIR::InstId prev_id,
-//        resolved instructions.
-//    - Build any necessary IR structures, and return the output constant.
-// 3. If resolve didn't return Retry(), pop the work. Otherwise, it needs to
-//    remain, and may no longer be at the top of the stack; set `retry` on it so
-//    we'll make sure to run it again later.
+//    - If HasNewWork() reports that any work was added, then return Retry():
+//      this instruction needs another call to complete. Gather the now-resolved
+//      constants and continue to the next step once the retry happens.
 //
-// TryResolveInst can complete in one call for a given instruction, but should
-// always complete within three calls:
+//    Second pass:
+//    - Build the constant value of the instruction.
+//    - Gather all input constants necessary to finish importing the
+//      instruction. This is only necessary for instructions like classes that
+//      can be forward-declared. For these instructions, we first import the
+//      constant value and then later import the rest of the declaration in
+//      order to break cycles.
+//    - If HasNewWork() reports that any work was added, then return
+//      Retry(constant_value): this instruction needs another call to complete.
+//      Gather the now-resolved constants and continue to the next step once the
+//      retry happens.
 //
-// - TryResolveInst can retry once if the generic is not yet loaded.
-// - TryResolveTypedInst can retry once if its inputs are not yet loaded.
-// - The third call should succeed.
+//    Third pass:
+//    - After this point, the constant value for the instruction is already set,
+//      and will be passed back into TryResolve*Inst on retry. It should not be
+//      created again.
+//    - Fill in any remaining information to complete the import of the
+//      instruction. For example, when importing a class declaration, build the
+//      class scope and information about the definition.
+//    - Return ResolveAs/ResolveAsConstant to finish the resolution process.
+//      This will cause the Resolve loop to set a constant value if we didn't
+//      retry at the end of the second pass.
+//
+// 3. If resolve didn't return Retry(), pop the work. Otherwise, it needs to
+//    remain, and may no longer be at the top of the stack; update the state of
+//    the work item to track what work still needs to be done.
 //
-// Due to the chance of a second call to TryResolveTypedInst, it's important to
-// reserve all expensive logic until it's been established that input constants
-// are available.
+// TryResolveInst can complete in one call for a given instruction. It should
+// always complete within three calls, although it is expected to be rare for
+// three calls to be needed. Due to the chance of a second or third call to
+// TryResolveInst, it's important to reserve all expensive logic until it's
+// been established that input constants are available.
```

Comment:
Had you considered moving this under step 2 docs, since that's mostly about TryResolveInst behavior?

Reply:
Done, merged into the suggested new text.